File: backend/src/models/DirectMessage.js
import mongoose from 'mongoose';

/**
 * A human-to-human message between a patient and the clinic (doctor or staff),
 * distinct from the AI assistant chat (ChatMessage). One flat thread per
 * patient — every message carries the patient it belongs to, so the patient's
 * inbox and each clinician's view are the same ordered conversation.
 */
const directMessageSchema = new mongoose.Schema(
  {
    patient: { type: mongoose.Schema.Types.ObjectId, ref: 'User', required: true, index: true },
    sender: { type: mongoose.Schema.Types.ObjectId, ref: 'User', required: true },
    // Who sent it, so the UI can align the bubble and label it ("Dr. Dey").
    senderRole: { type: String, enum: ['patient', 'doctor', 'staff'], required: true },

    content: { type: String, required: true, trim: true, maxlength: 4000 },

    // Read receipts, one per side, so each can show an unread badge.
    readByPatient: { type: Boolean, default: false },
    readByClinician: { type: Boolean, default: false },
  },
  { timestamps: true },
);

directMessageSchema.index({ patient: 1, createdAt: 1 });

export const DirectMessage = mongoose.model('DirectMessage', directMessageSchema);

File: mobile/lib/features/messaging/domain/direct_message.dart
/// One human message between a patient and the clinic (not the AI assistant).
class DirectMessage {
  const DirectMessage({
    required this.id,
    required this.senderRole,
    required this.content,
    this.senderName,
    this.createdAt,
  });

  final String id;

  /// 'patient' | 'doctor' | 'staff'
  final String senderRole;
  final String content;
  final String? senderName;
  final DateTime? createdAt;

  bool get fromPatient => senderRole == 'patient';

  factory DirectMessage.fromJson(Map<String, dynamic> j) => DirectMessage(
    id: j['id']?.toString() ?? '',
    senderRole: j['senderRole']?.toString() ?? 'patient',
    content: j['content']?.toString() ?? '',
    senderName: j['senderName']?.toString(),
    createdAt: DateTime.tryParse(j['createdAt']?.toString() ?? '')?.toLocal(),
  );
}

/// A row in the clinician's message inbox — one conversation per patient.
class MessageThread {
  const MessageThread({
    required this.patientId,
    this.patientName,
    this.patientPhone,
    this.lastMessage,
    this.lastAt,
    this.lastSenderRole,
    this.unread = 0,
  });

  final String patientId;
  final String? patientName;
  final String? patientPhone;
  final String? lastMessage;
  final DateTime? lastAt;
  final String? lastSenderRole;
  final int unread;

  factory MessageThread.fromJson(Map<String, dynamic> j) => MessageThread(
    patientId: j['patientId']?.toString() ?? '',
    patientName: j['patientName']?.toString(),
    patientPhone: j['patientPhone']?.toString(),
    lastMessage: j['lastMessage']?.toString(),
    lastAt: DateTime.tryParse(j['lastAt']?.toString() ?? '')?.toLocal(),
    lastSenderRole: j['lastSenderRole']?.toString(),
    unread: (j['unread'] as num?)?.toInt() ?? 0,
  );
}
